File: src/email/aliases.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::error::Error;

#[derive(Clone)]
pub struct AliasClient {
    client: Client,
    base_url: String,
    api_key: Option<String>,
}

#[derive(Deserialize, Debug)]
pub struct Mailbox {
    pub id: u32,
    pub email: String,
}

#[derive(Deserialize, Debug)]
pub struct Alias {
    pub id: u32,
    pub email: String,
    pub enabled: bool,
    pub mailboxes: Vec<Mailbox>,
    pub name: Option<String>,
}

#[derive(Deserialize, Debug)]
pub struct AliasesResponse {
    pub aliases: Vec<Alias>,
}

#[derive(Serialize)]
struct CreateAliasRequest<'a> {
    alias_prefix: &'a str,
    signed_suffix: &'a str,
    mailbox_ids: Vec<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    note: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    name: Option<&'a str>,
}


#[derive(Serialize)]
struct CreateRandomAliasRequest<'a> {
    #[serde(skip_serializing_if = "Option::is_none")]
    note: Option<&'a str>,
}

impl AliasClient {
    pub fn new(client: Client, base_url: String, api_key: Option<String>) -> Self {
        Self {
            client,
            base_url,
            api_key,
        }
    }

    pub async fn list_aliases(&self, page_id: u32) -> Result<AliasesResponse, Box<dyn Error>> {
        let api_key = self.api_key.as_ref().ok_or("API Key not set")?;
        let res = self
            .client
            .get(format!(
                "{}/api/v2/aliases?page_id={}",
                self.base_url, page_id
            ))
            .header("Authentication", api_key)
            .send()
            .await?;

        if res.status().is_success() {
            Ok(res.json::<AliasesResponse>().await?)
        } else {
            Err(format!("Failed to list aliases: {}", res.text().await?).into())
        }
    }

    pub async fn create_alias(
        &self,
        alias_prefix: &str,
        signed_suffix: &str,
        mailbox_ids: Vec<u32>,
        note: Option<&str>,
        name: Option<&str>,
    ) -> Result<Alias, Box<dyn Error>> {
        let api_key = self.api_key.as_ref().ok_or("API Key not set")?;
        let req_body = CreateAliasRequest { alias_prefix, signed_suffix, mailbox_ids, note, name };
        let res = self.client
            .post(format!("{}/api/v3/alias/custom/new", self.base_url))
            .header("Authentication",  api_key)
            .json(&req_body)
            .send()
            .await?;

        match res.status() {
            reqwest::StatusCode::CREATED => Ok(res.json::<Alias>().await?),
            _ => Err(format!("Failed to create alias: {}", res.text().await?).into()),
        }
    }

    /// Create a new random alias
    pub async fn create_random_alias(
        &self,
        hostname: Option<&str>,
        mode: Option<&str>,  // "uuid" or "word"
        note: Option<&str>,
    ) -> Result<Alias, Box<dyn Error>> {
        let api_key = self.api_key.as_ref().ok_or("API Key not set")?;

        let mut url = format!("{}/api/alias/random/new", self.base_url);
        let mut query_params = vec![];
        if let Some(host) = hostname {
            query_params.push(format!("hostname={}", host));
        }
        if let Some(m) = mode {
            query_params.push(format!("mode={}", m));
        }
        if !query_params.is_empty() {
            url.push('?');
            url.push_str(&query_params.join("&"));
        }

        let req_body = CreateRandomAliasRequest { note };

        let res = self.client
            .post(url)
            .header("Authentication", api_key)
            .json(&req_body)
            .send()
            .await?;

        match res.status() {
            reqwest::StatusCode::CREATED => Ok(res.json::<Alias>().await?),
            _ => Err(format!("Failed to create random alias: {}", res.text().await?).into()),
        }
    }

    /// Delete an alias
    pub async fn delete_alias(&self, alias_id: u32) -> Result<bool, Box<dyn Error>> {
        let api_key = self.api_key.as_ref().ok_or("API Key not set")?;
        let res = self.client
            .delete(format!("{}/api/aliases/{}", self.base_url, alias_id))
            .header("Authentication", api_key)
            .send()
            .await?;

        match res.status() {
            reqwest::StatusCode::OK => Ok(true),
            _ => Err(format!("Failed to delete alias: {}", res.text().await?).into()),
        }
    }
}

File: vault/email_aliases/aliases.py
import httpx
from pydantic import BaseModel
from typing import Optional, List


# ---------------------------------------------------------------------------
# Data Models
# ---------------------------------------------------------------------------

class Mailbox(BaseModel):
    id: int
    email: str


class Alias(BaseModel):
    id: int
    email: str
    enabled: bool
    mailboxes: List[Mailbox]
    name: Optional[str] = None


class AliasesResponse(BaseModel):
    aliases: List[Alias]


class CreateAliasRequest(BaseModel):
    alias_prefix: str
    signed_suffix: str
    mailbox_ids: List[int]
    note: Optional[str] = None
    name: Optional[str] = None


class CreateRandomAliasRequest(BaseModel):
    note: Optional[str] = None


# ---------------------------------------------------------------------------
# Client Implementation
# ---------------------------------------------------------------------------

class AliasClient:
    def __init__(self, base_url: str, api_key: Optional[str] = None, client: Optional[httpx.AsyncClient] = None):
        """
        :param base_url: The API base URL.
        :param api_key: Optional API key for authentication.
        :param client: Optional existing httpx.AsyncClient.
        """
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        # Use provided client or create a new one
        self.client = client if client else httpx.AsyncClient()

    async def list_aliases(self, page_id: int) -> AliasesResponse:
        """
        Fetches a list of aliases (paginated).
        """
        if not self.api_key:
            raise ValueError("API Key not set")

        url = f"{self.base_url}/api/v2/aliases"
        params = {"page_id": page_id}
        headers = {"Authentication": self.api_key}

        try:
            response = await self.client.get(url, params=params, headers=headers)

            if response.status_code == 200:
                return AliasesResponse(**response.json())
            else:
                raise Exception(f"Failed to list aliases: {response.text}")

        except httpx.RequestError as e:
            raise Exception(f"Network error occurred: {e}")

    async def create_alias(
            self,
            alias_prefix: str,
            signed_suffix: str,
            mailbox_ids: List[int],
            note: Optional[str] = None,
            name: Optional[str] = None,
    ) -> Alias:
        """
        Creates a new custom alias.
        """
        if not self.api_key:
            raise ValueError("API Key not set")

        url = f"{self.base_url}/api/v3/alias/custom/new"
        headers = {"Authentication": self.api_key}

        payload = CreateAliasRequest(
            alias_prefix=alias_prefix,
            signed_suffix=signed_suffix,
            mailbox_ids=mailbox_ids,
            note=note,
            name=name
        )

        # use exclude_none=True to mimic Rust's skip_serializing_if = "Option::is_none"
        json_body = payload.model_dump(exclude_none=True)

        try:
            response = await self.client.post(url, headers=headers, json=json_body)

            if response.status_code == 201:
                return Alias(**response.json())
            else:
                raise Exception(f"Failed to create alias: {response.text}")

        except httpx.RequestError as e:
            raise Exception(f"Network error occurred: {e}")

    async def create_random_alias(
            self,
            hostname: Optional[str] = None,
            mode: Optional[str] = None,  # "uuid" or "word"
            note: Optional[str] = None,
    ) -> Alias:
        """
        Creates a new random alias.
        """
        if not self.api_key:
            raise ValueError("API Key not set")

        url = f"{self.base_url}/api/alias/random/new"
        headers = {"Authentication": self.api_key}

        # Build query parameters
        params = {}
        if hostname:
            params["hostname"] = hostname
        if mode:
            params["mode"] = mode

        payload = CreateRandomAliasRequest(note=note)
        json_body = payload.model_dump(exclude_none=True)

        try:
            response = await self.client.post(url, headers=headers, params=params, json=json_body)

            if response.status_code == 201:
                return Alias(**response.json())
            else:
                raise Exception(f"Failed to create random alias: {response.text}")

        except httpx.RequestError as e:
            raise Exception(f"Network error occurred: {e}")

    async def delete_alias(self, alias_id: int) -> bool:
        """
        Deletes an alias by ID.
        """
        if not self.api_key:
            raise ValueError("API Key not set")

        url = f"{self.base_url}/api/aliases/{alias_id}"
        headers = {"Authentication": self.api_key}

        try:
            response = await self.client.delete(url, headers=headers)

            if response.status_code == 200:
                return True
            else:
                raise Exception(f"Failed to delete alias: {response.text}")

        except httpx.RequestError as e:
            raise Exception(f"Network error occurred: {e}")

    async def close(self):
        """Closes the underlying HTTP client."""
        await self.client.aclose()